lib/roar/json/json_api/declarative.rb

```diff
@@ -0,0 +1,100 @@
+module Roar
+  module JSON
+    module JSONAPI
+      # New API for JSON-API representers.
+      module Declarative
+        def type(name = nil)
+          return @type unless name # original name.
+
+          heritage.record(:type, name)
+          @type = name.to_s
+        end
+
+        def link(name, options = {}, &block)
+          return super(name, &block) unless options[:toplevel]
+          for_collection.link(name, &block)
+        end
+
+        def meta(options = {}, &block)
+          return super(&block) unless options[:toplevel]
+          for_collection.meta(&block)
+        end
+
+        def relationship(options = {}, &block); end
+
+        def has_one(name, options = {}, &block)
+          has_relationship(name, options.merge(collection: false), &block)
+        end
+
+        def has_many(name, options = {}, &block)
+          has_relationship(name, options.merge(collection: true), &block)
+        end
+
+        private
+
+        def has_relationship(name, options = {}, &block)
+          # every nested representer is a full-blown JSONAPI representer.
+          nested(:included, inherit: true) do
+            property(name, collection: options[:collection]) {
+              include Roar::JSON::JSONAPI
+
+              instance_exec(&block)
+
+              def from_document(hash)
+                hash
+              end
+            }
+          end
+
+          resource_identifier_representer = Class.new(Roar::Decorator)
+          resource_identifier_representer.class_eval do
+            include Roar::JSON
+            include Roar::Hypermedia
+            include JSONAPI::Meta
+            extend JSONAPI::Declarative
+
+            def self.relationship_block
+              @relationship_block || -> {}
```

Comment:
A JSON API `relationship` has special semantics, hence my preference for a `relationship` block. In fact, reading the spec again, it appears even empty to-one relationships can have links/meta associated with them.

However, let's suppose we want to implement this in a similar way to the way `toplevel: true` is implemented:

```ruby
  link(:self, :relationship: true)     { "/articles/#{represented.id}/relationships/comments" }
  link(:related, :relationship: true)  { "/articles/#{represented.id}/comments" }
  meta :relationship: true do
    property :count
  end
```

This is predicated on the `::for_collection` API:

```ruby
        def link(name, options = {}, &block)
          return super(name, &block) unless options[:toplevel]
          for_collection.link(name, &block)
        end
```

AFAICT there is no built-in mechanism for a [child representer](https://github.com/trailblazer/roar-jsonapi/pull/16/files#diff-5cf99150a8f672cf05cc67c7a054c120R80) to keep a reference to [its parent](https://github.com/trailblazer/roar-jsonapi/pull/16/files#diff-5cf99150a8f672cf05cc67c7a054c120R74).

```ruby
        def link(name, options = {}, &block)
          return for_collection.link(name, &block) if options[:toplevel]
          return for_relationship.link(name, &block) if options[:relationship] # ref to parent
          super(name, &block)
        end
```

Or am I missing something?